packages/@lwc/engine/src/framework/hot-swaps.ts

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { VM, scheduleRehydration } from './vm';
+import { isFalse, isUndefined, isNull } from '@lwc/shared';
+import { markComponentAsDirty, ComponentConstructor } from './component';
+import { Template } from './template';
+import { StylesheetFactory } from './stylesheet';
+
+const swappedTemplateMap = new WeakMap<Template, Template>();
+const swappedComponentMap = new WeakMap<ComponentConstructor, ComponentConstructor>();
+const swappedStyleMap = new WeakMap<StylesheetFactory, StylesheetFactory>();
+
+const activeTemplates = new WeakMap<Template, Set<VM>>();
+const activeComponents = new WeakMap<ComponentConstructor, Set<VM>>();
+const activeStyles = new WeakMap<StylesheetFactory, Set<VM>>();
+
+function rehydrateHotTemplate(tpl: Template): boolean {
+    const list = activeTemplates.get(tpl);
+    list?.forEach((vm) => {
```

Comment:
Are we sure list will never be undefined here? Consider a case where a user will create a component and call `swapTemplate` before the components initial hydration. This statement will fail.
```js
const element = createElement('x-foo, { is: MyFoo });
swapTemplate(initialTemplate, newTemplate);
```

Reply:
I don't think it will fail... the map in the template will be added, but not rehydrate will be issued at that point because no one has used that template yet. But when the time comes to render MyFoo instance, it will attempt to return the old template when render() is called, and it will then be resolved to  the new template. That's how it works.

As for the list being undefined, yes, that's what `?.` does there.